### ISSUE TITLE
long paths are very slow

### ISSUE THREAD

**Issue by JeffBezanson** [nolta/Winston.jl#89]
If I'm understanding this right it's really a performance bug in Cairo, but drawing very long paths is O(n^2). This was discussed on the mailing list a while back. Just messing around, I tried the following Winston patch:

```
diff --git a/src/renderer.jl b/src/renderer.jl
index 029ff93..a6e4c0c 100644
--- a/src/renderer.jl
+++ b/src/renderer.jl
@@ -245,9 +245,12 @@ function curve(self::CairoRenderer, x::AbstractVector, y::A
         move_to(self.ctx, x[i], y[i])
         for i = i+1:n
             line_to(self, x[i], y[i])
+            if i == n || i%32==0
+                stroke(self.ctx)
+                move_to(self, x[i], y[i])
+            end
         end
     end
-    stroke(self.ctx)
 end
```

This makes plotting long paths _insanely_ faster for me. @timholy I recall there was some other workaround for this, but perhaps we should do something like this inside Winston to make sure people don't hit the slow case?

**Comment by stuaxo** [comment]
Hi,
   I just came across this ticket randomly.   Do you know if there is a cairo bug for this ?
I don't know Julia, if I were to try and reproduce this with pycairo or Cairo in C, is it just a matter of creating a path made up of a very large number of straight lines ?

Cheers
Stuart

**Comment by lobingera** [comment]
@stuaxo Hi, i saw your post on the Cairo mailing list. What problem are you trying to solve?

**Comment by stuaxo** [comment]
Hi,   I'm a developer on Shoebot, which presents users with a vector drawing API backed by Cairo.
Users can draw anything they like, from simple to complex shapes.
If the default install of Cairo is slow at drawing long paths vs drawing the same path made of lots of single segment paths, then I'd like to collect info and submit a bug.
TLDR:  I'm interested in Cairo getting faster and this bug looks like it may point to an optimization for Cairo itself.

S++ 

    On Wednesday, May 26, 2021, 4:34:54 PM GMT+1, Andreas Lobinger ***@***.***> wrote:  
 
 


@stuaxo Hi, i saw your post on the Cairo mailing list. What problem are you trying to solve?

**Comment by stuaxo** [comment]
Cheers - I've replied on the mailinglist, since it's a bit off topic there.

The answer was I'm not trying to do anything in particular but I'd like a quicker Cairo in general.

**Comment by lobingera** [comment]
https://github.com/servo/pathfinder

**Comment by stuaxo** [comment]
Thanks, I'm aware of pathfinder - but I don't think someone has made python bindings for it and wrapped it in a WHL yet. 

It would also mean rewriting our 10 year old Cairo/Pango based codebase, but that's not as important as the first point.

Shoebot the project has taught me that if a user has to install any dependency using something other than the languages package manager, they appear on our bug tracker, and it's often difficult get them up and running.

This was the case with Pycairo, which has a python package, if not a always a WHL (for non WHL versions need the user to have a compiler setup, which they often do not on non-linux platforms).

When we do eventually rewrite, I'm will only choose dependencies that have WHLs on all 3 major platforms: if users can't a library, then it may as well not exist.
